### PR TITLE
PARQUET-711: Use metadata builders in parquet writer

### PR DESCRIPTION
**Pull request by majetideepak**
I wrote a sample file and the metadata seems to be correct.
@xhochy I fixed some missing metadata like `dictionary_page_offset`. You might want to check if this fixes the Drill problem.

### CODE REVIEW

File: src/parquet/file/writer-internal.cc [hunk 1]
```diff
@@ -133,50 +133,38 @@ int64_t SerializedPageWriter::WriteDictionaryPage(const DictionaryPage& page) {
...
 
     if (current_column_writer_) {
-      total_bytes_written_ += current_column_writer_->Close();
+      total_bytes_written_ = current_column_writer_->Close();
       current_column_writer_.reset();
     }
-
-    metadata_->__set_total_byte_size(total_bytes_written_);
+    // Ensures all columns have been written
+    metadata_->Finish(total_bytes_written_);
```
Comment (majetideepak) [inline src/parquet/file/writer-internal.cc]: `total_bytes_written_` value updated by the `ColumnWriter` is actually redundant. `total_bytes_written_` is computed in the metadata builder by adding the `compressed_size` of all columns. 
I am still passing the value here for checking correctness. There is a DCHECK inside `RowGroupMetaDataBuilder::Finish()`.

File: src/parquet/column/column-io-benchmark.cc [hunk 2]
```diff
@@ -57,17 +57,19 @@ void SetBytesProcessed(::benchmark::State& state, Repetition::type repetition) {
...
 template <Repetition::type repetition>
 static void BM_WriteInt64Column(::benchmark::State& state) {
-  format::ColumnChunk metadata;
+  format::ColumnChunk thrift_metadata;
   std::vector<int64_t> values(state.range_x(), 128);
   std::vector<int16_t> definition_levels(state.range_x(), 1);
   std::vector<int16_t> repetition_levels(state.range_x(), 0);
   std::shared_ptr<ColumnDescriptor> schema = Int64Schema(repetition);
   std::shared_ptr<parquet::WriterProperties> properties = default_writer_properties();
+  auto metadata = parquet::ColumnChunkMetaDataBuilder::Make(
```
Comment (xhochy) [inline src/parquet/column/column-io-benchmark.cc]: No need for the `parquet::` prefix here, the benchmarks are also in the same namespace.

File: src/parquet/file/writer.h [hunk 3]
```diff
@@ -152,19 +143,9 @@ class PARQUET_EXPORT ParquetFileWriter {
    */
   const std::shared_ptr<WriterProperties>& properties() const;
 
-  /**
-   * Returns the file schema descriptor
-   */
-  const SchemaDescriptor* descr() { return schema_; }
```
Comment (xhochy) [inline src/parquet/file/writer.h]: These two function are quite comfortable, I'd like to preserve them. It's ok for me if they call a bit deeper but they are just convenient from an API user level.

**Comment by xhochy** [comment]
@majetideepak Perfect! This fixes the Drill issue.

**Comment by xhochy** [comment]
Except for my wish to keep the two helper functions: LGTM

**Comment by majetideepak** [comment]
I will add those two helper functions back

File: src/parquet/file/writer-internal.h [hunk 4]
```diff
@@ -67,24 +71,17 @@ class SerializedPageWriter : public PageWriter {
...
-  RowGroupSerializer(int64_t num_rows, const SchemaDescriptor* schema, OutputStream* sink,
-      format::RowGroup* metadata, const WriterProperties* properties)
+  RowGroupSerializer(int64_t num_rows, OutputStream* sink,
+      RowGroupMetaDataBuilder* metadata, const WriterProperties* properties)
       : num_rows_(num_rows),
-        schema_(schema),
         sink_(sink),
-        metadata_(metadata),
-        allocator_(properties->allocator()),
+        metadata_(std::move(metadata)),
```
Comment (wesm) [inline src/parquet/file/writer-internal.h]: This move not needed?

File: src/parquet/file/writer.cc [hunk 5]
```diff
@@ -64,9 +61,16 @@ std::unique_ptr<ParquetFileWriter> ParquetFileWriter::Open(
   return result;
 }
 
+const SchemaDescriptor* ParquetFileWriter::descr() const {
```
Comment (wesm) [inline src/parquet/file/writer.cc]: Maybe call this `schema()`?

**Comment by wesm** [comment]
I admit when to use `descr` or `schema` names is unclear. We should not use `schema` to refer to the `GroupNode` instance that initializes the `SchemaDescriptor`, and perhaps then `schema` should only refer to the `SchemaDescriptor` object (and use `descr` when referring to `ColumnDescriptor`). Doesn't need to be all handled in this patch.

**Comment by majetideepak** [comment]
Let's use `schema` for `SchemaDescriptor`, `descr` for `ColumnDescriptor` and `schema_node` for the `GroupNode`. I will implement PARQUET-705 and PARQUET-707 in this patch

File: src/parquet/schema/descriptor.h [hunk 6]
```diff
@@ -100,18 +100,20 @@ class PARQUET_EXPORT SchemaDescriptor {
...
   const schema::NodePtr& GetColumnRoot(int i) const;
 
+  const std::string& name() const { return group_node_->name(); }
+
  private:
   friend class ColumnDescriptor;
 
   schema::NodePtr schema_;
-  const schema::GroupNode* group_;
+  const schema::GroupNode* group_node_;
```
Comment (majetideepak) [inline src/parquet/schema/descriptor.h]: This is consistent with `primitive_node_` used in the `ColumnDescriptor`

File: src/parquet/schema/descriptor.h [hunk 7]
```diff
@@ -100,18 +100,20 @@ class PARQUET_EXPORT SchemaDescriptor {
...
-  const schema::NodePtr& schema() const { return schema_; }
+  const schema::NodePtr& schema_root() const { return schema_; }
 
-  const schema::GroupNode* group() const { return group_; }
+  const schema::GroupNode* group_node() const { return group_node_; }
 
   // Returns the root (child of the schema root) node of the leaf(column) node
   const schema::NodePtr& GetColumnRoot(int i) const;
 
+  const std::string& name() const { return group_node_->name(); }
```
Comment (majetideepak) [inline src/parquet/schema/descriptor.h]: PARQUET-707

**Comment by wesm** [comment]
+1, thank you!

**Pull request closed without merging**